This Notebook will be used for capstone project.

<H1> PART -I

Scrapping the toronto neighbourhood data from wiki using beautiful soup

In [2]:
import requests
from bs4 import BeautifulSoup

url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
page = requests.get(url)
soup = BeautifulSoup(page.content, 'html.parser')
tb = soup.find('table', class_='wikitable')
table_tmp = tb.tbody.find_all("tr")


In [3]:
headings = []
for th in table_tmp[0].find_all("th"):
    # remove any newlines and extra spaces from left and right
    headings.append(th.text.replace('\n', ' ').strip())

print(headings)


['Postcode', 'Borough', 'Neighbourhood']


We got the headings now we can opaerse data in dictionary and then convert it to data frame

In [4]:
table_data = []
for tr in tb.tbody.find_all("tr"): # find all tr's from table's tbody
        t_row = {}
        
        for td, th in zip(tr.find_all("td"), headings): 
            t_row[th] = td.text.replace('\n', '').strip()
        table_data.append(t_row)


In [5]:
postcode = []
borough =[] 
neighbourhood = []

for item in table_data[1:]:
    postcode.append(item['Postcode'])
    borough.append(item['Borough'])
    neighbourhood.append(item['Neighbourhood'])
    

converted to 3 seprate list checked each list has same length.

In [6]:
print(neighbourhood[:5] , postcode[:5] , borough[:5])

['Not assigned', 'Not assigned', 'Parkwoods', 'Victoria Village', 'Harbourfront'] ['M1A', 'M2A', 'M3A', 'M4A', 'M5A'] ['Not assigned', 'Not assigned', 'North York', 'North York', 'Downtown Toronto']


In [7]:
print(len(neighbourhood),len(postcode),len(borough))

287 287 287


In [8]:
import pandas as pd

In [9]:
df = pd.DataFrame({"PostalCode" : postcode,
              "Borough" :borough,
              "Neighborhood" :neighbourhood})

In [10]:
df.head()

PostalCode           Borough      Neighborhood
0        M1A      Not assigned      Not assigned
1        M2A      Not assigned      Not assigned
2        M3A        North York         Parkwoods
3        M4A        North York  Victoria Village
4        M5A  Downtown Toronto      Harbourfront

ignoring the not assigned boroughs

In [11]:
df = df[df.Borough != "Not assigned"]

In [12]:
df.shape

(210, 3)

In [13]:
df.head()

PostalCode           Borough      Neighborhood
2        M3A        North York         Parkwoods
3        M4A        North York  Victoria Village
4        M5A  Downtown Toronto      Harbourfront
5        M6A        North York  Lawrence Heights
6        M6A        North York    Lawrence Manor

Finding the rows that have neighboehood not assigned

In [14]:
df[df.Neighborhood ==  "Not assigned"]


PostalCode       Borough  Neighborhood
7        M7A  Queen's Park  Not assigned

Found which rows have not assigned neighborhood

Converting those non assigned neighborhood values to default vales of borough

In [15]:
df.loc[df.Neighborhood ==  "Not assigned","Neighborhood"] = df["Borough"]

In [16]:
df.head(10)

PostalCode           Borough      Neighborhood
2         M3A        North York         Parkwoods
3         M4A        North York  Victoria Village
4         M5A  Downtown Toronto      Harbourfront
5         M6A        North York  Lawrence Heights
6         M6A        North York    Lawrence Manor
7         M7A      Queen's Park      Queen's Park
9         M9A      Queen's Park      Queen's Park
10        M1B       Scarborough             Rouge
11        M1B       Scarborough           Malvern
13        M3B        North York   Don Mills North

Grouped unique postal codes with neighbourhood.

In [17]:
df = df.groupby(['PostalCode','Borough'])['Neighborhood'].apply(lambda x: ','.join(x.unique())).reset_index()

In [18]:
df.head(20)

PostalCode      Borough                                       Neighborhood
0         M1B  Scarborough                                      Rouge,Malvern
1         M1C  Scarborough               Highland Creek,Rouge Hill,Port Union
2         M1E  Scarborough                    Guildwood,Morningside,West Hill
3         M1G  Scarborough                                             Woburn
4         M1H  Scarborough                                          Cedarbrae
5         M1J  Scarborough                                Scarborough Village
6         M1K  Scarborough          East Birchmount Park,Ionview,Kennedy Park
7         M1L  Scarborough                      Clairlea,Golden Mile,Oakridge
8         M1M  Scarborough      Cliffcrest,Cliffside,Scarborough Village West
9         M1N  Scarborough                         Birch Cliff,Cliffside West
10        M1P  Scarborough  Dorset Park,Scarborough Town Centre,Wexford He...
11        M1R  Scarborough                                   Maryvale,Wexford
12        M1S  Scarborough                                          Agincourt
13        M1T  Scarborough              Clarks Corners,Sullivan,Tam O'Shanter
14        M1V  Scarborough  Agincourt North,L'Amoreaux East,Milliken,Steel...
15        M1W  Scarborough                                    L'Amoreaux West
16        M1X  Scarborough                                        Upper Rouge
17        M2H   North York                                  Hillcrest Village
18        M2J   North York                         Fairview,Henry Farm,Oriole
19        M2K   North York                                    Bayview Village

In [19]:
df.shape

(103, 3)

We have got 103 rows and 3 columns 

## PART -2 

Get the latitude and longitude for each neighborhood

In [20]:
#!pip install geocoder

In [21]:
#import geocoder # import geocoder

# initialize your variable to None
#lat_lng_coords = None

# loop until you get the coordinates
#while(lat_lng_coords is None):
#  g = geocoder.google('{}, Toronto, Ontario'.format('M1B'))
#  lat_lng_coords = g.latlng

#latitude = lat_lng_coords[0]
#longitude = lat_lng_coords[1]

In [22]:
#g = geocoder.google('Mountain View, CA')
#g.latlng


tried executing geocoder but it keeps running forever and dosent display any output. So using the csv file provided 

In [23]:
df1 = pd.read_csv("D:/Downloads/Geospatial_Coordinates.csv")

In [25]:
df1.head()

Postal Code   Latitude  Longitude
0         M1B  43.806686 -79.194353
1         M1C  43.784535 -79.160497
2         M1E  43.763573 -79.188711
3         M1G  43.770992 -79.216917
4         M1H  43.773136 -79.239476

we have to merge this both datasets

In [29]:
df_merged = pd.merge(left= df , right= df1, left_on = "PostalCode",right_on = "Postal Code")

In [30]:
df_merged.head()

PostalCode      Borough                          Neighborhood Postal Code  \
0        M1B  Scarborough                         Rouge,Malvern         M1B   
1        M1C  Scarborough  Highland Creek,Rouge Hill,Port Union         M1C   
2        M1E  Scarborough       Guildwood,Morningside,West Hill         M1E   
3        M1G  Scarborough                                Woburn         M1G   
4        M1H  Scarborough                             Cedarbrae         M1H   

    Latitude  Longitude  
0  43.806686 -79.194353  
1  43.784535 -79.160497  
2  43.763573 -79.188711  
3  43.770992 -79.216917  
4  43.773136 -79.239476

Dropping the extra Postal Code column

In [33]:
df_merged = df_merged.drop("Postal Code",axis =1)

In [34]:
df_merged.head()

PostalCode      Borough                          Neighborhood   Latitude  \
0        M1B  Scarborough                         Rouge,Malvern  43.806686   
1        M1C  Scarborough  Highland Creek,Rouge Hill,Port Union  43.784535   
2        M1E  Scarborough       Guildwood,Morningside,West Hill  43.763573   
3        M1G  Scarborough                                Woburn  43.770992   
4        M1H  Scarborough                             Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

## Part - 3 

Analysisng toronto neighborhood using foursquare api 

In [38]:
#!pip install geopy

twisted 18.7.0 requires PyHamcrest>=1.9.0, which is not installed.
You are using pip version 10.0.1, however version 19.3.1 is available.
You should consider upgrading via the 'python -m pip install --upgrade pip' command.


In [39]:
import numpy as np # library to handle data in a vectorized manner

pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

import json # library to handle JSON files

from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests
from pandas.io.json import json_normalize # tranform JSON file into a pandas dataframe

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

import folium # map rendering library

print('Libraries imported.')

Libraries imported.


In [41]:
address = 'Toronto, ON'

geolocator = Nominatim(user_agent="tn_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


In [44]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=10)

# add markers to map
for lat, lng, borough, neighborhood in zip(df_merged['Latitude'], df_merged['Longitude'], df_merged['Borough'], df_merged['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7).add_to(map_toronto)  
    
map_toronto

To make our analysis simplified we will just work on those borough that have toronto name in it!

In [49]:
toronto_data = df_merged[df_merged['Borough'].str.contains('Toronto', regex=False)].reset_index(drop=True)
toronto_data.head()

PostalCode          Borough                   Neighborhood   Latitude  \
0        M4E     East Toronto                    The Beaches  43.676357   
1        M4K     East Toronto    The Danforth West,Riverdale  43.679557   
2        M4L     East Toronto  The Beaches West,India Bazaar  43.668999   
3        M4M     East Toronto                Studio District  43.659526   
4        M4N  Central Toronto                  Lawrence Park  43.728020   

   Longitude  
0 -79.293031  
1 -79.352188  
2 -79.315572  
3 -79.340923  
4 -79.388790

In [51]:
toronto_data["Borough"].unique()

array(['East Toronto', 'Central Toronto', 'Downtown Toronto',
       'West Toronto'], dtype=object)

Since our data is just focused around Central Toronto getting coordinates using Foursquare api for this location

In [54]:
address = 'Central Toronto,Toronto'

geolocator = Nominatim(user_agent="tn_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Toronto are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Toronto are 43.653963, -79.387207.


In [58]:
# create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# add markers to map
for lat, lng, borough, neighborhood in zip(toronto_data['Latitude'], toronto_data['Longitude'], toronto_data['Borough'], toronto_data['Neighborhood']):
    label = '{}, {}'.format(neighborhood, borough)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7).add_to(map_toronto)  
    
map_toronto

#### Define Foursquare Credentials and Version

In [59]:
CLIENT_ID = '2F0YVH52B53OB0MCMNR3SXCQBRKYNOC3YDJ23NJQXKXY5EFR' # your Foursquare ID
CLIENT_SECRET = 'QLYK2YW4BTWD1PBCWTOIMHNIGL0JC5X14G0APKAR1FR5T4AG' # your Foursquare Secret
VERSION = '20191031' # Foursquare API version taking last month api for analysis

print('Your credentails:')
print('CLIENT_ID: ' + CLIENT_ID)
print('CLIENT_SECRET:' + CLIENT_SECRET)

Your credentails:
CLIENT_ID: 2F0YVH52B53OB0MCMNR3SXCQBRKYNOC3YDJ23NJQXKXY5EFR
CLIENT_SECRET:QLYK2YW4BTWD1PBCWTOIMHNIGL0JC5X14G0APKAR1FR5T4AG


## 2. Explore Neighborhoods in Toronto


Lets get the top 100 venues in this area with their categories within 500m radius

In [60]:
LIMIT = 100
def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [61]:
toronto_venues = getNearbyVenues(names=toronto_data['Neighborhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude']
                                  )


The Beaches
The Danforth West,Riverdale
The Beaches West,India Bazaar
Studio District
Lawrence Park
Davisville North
North Toronto West
Davisville
Moore Park,Summerhill East
Deer Park,Forest Hill SE,Rathnelly,South Hill,Summerhill West
Rosedale
Cabbagetown,St. James Town
Church and Wellesley
Harbourfront
Ryerson,Garden District
St. James Town
Berczy Park
Central Bay Street
Adelaide,King,Richmond
Harbourfront East,Toronto Islands,Union Station
Design Exchange,Toronto Dominion Centre
Commerce Court,Victoria Hotel
Roselawn
Forest Hill North,Forest Hill West
The Annex,North Midtown,Yorkville
Harbord,University of Toronto
Chinatown,Grange Park,Kensington Market
CN Tower,Bathurst Quay,Island airport,Harbourfront West,King and Spadina,Railway Lands,South Niagara
Stn A PO Boxes 25 The Esplanade
First Canadian Place,Underground city
Christie
Dovercourt Village,Dufferin
Little Portugal,Trinity
Brockton,Exhibition Place,Parkdale Village
High Park,The Junction South
Parkdale,Roncesvalles
Runnymede

#### Let's check the size of the resulting dataframe

In [62]:
print(toronto_venues.shape)
toronto_venues.head()

(1698, 7)


Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0                  The Beaches              43.676357              -79.293031   
1                  The Beaches              43.676357              -79.293031   
2                  The Beaches              43.676357              -79.293031   
3                  The Beaches              43.676357              -79.293031   
4  The Danforth West,Riverdale              43.679557              -79.352188   

                                Venue  Venue Latitude  Venue Longitude  \
0                   Glen Manor Ravine       43.676821       -79.293942   
1  The Big Carrot Natural Food Market       43.678879       -79.297734   
2                 Grover Pub and Grub       43.679181       -79.297215   
3                       Upper Beaches       43.680563       -79.292869   
4                            Pantheon       43.677621       -79.351434   

      Venue Category  
0              Trail  
1  Health Food Store  
2                Pub  
3       Neighborhood  
4   Greek Restaurant

Let's check how many venues were returned for each neighborhood

In [63]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Adelaide,King,Richmond                                                100   
Berczy Park                                                            58   
Brockton,Exhibition Place,Parkdale Village                             22   
Business Reply Mail Processing Centre 969 Eastern                      18   
CN Tower,Bathurst Quay,Island airport,Harbourfr...                     16   
Cabbagetown,St. James Town                                             44   
Central Bay Street                                                     83   
Chinatown,Grange Park,Kensington Market                                95   
Christie                                                               17   
Church and Wellesley                                                   85   
Commerce Court,Victoria Hotel                                         100   
Davisville                                                             35   
Davisville North                                                        8   
Deer Park,Forest Hill SE,Rathnelly,South Hill,S...                     16   
Design Exchange,Toronto Dominion Centre                               100   
Dovercourt Village,Dufferin                                            15   
First Canadian Place,Underground city                                 100   
Forest Hill North,Forest Hill West                                      5   
Harbord,University of Toronto                                          35   
Harbourfront                                                           48   
Harbourfront East,Toronto Islands,Union Station                       100   
High Park,The Junction South                                           25   
Lawrence Park                                                           3   
Little Portugal,Trinity                                                64   
Moore Park,Summerhill East                                              4   
North Toronto West                                                     21   
Parkdale,Roncesvalles                                                  15   
Rosedale                                                                4   
Roselawn                                                                1   
Runnymede,Swansea                                                      37   
Ryerson,Garden District                                               100   
St. James Town                                                        100   
Stn A PO Boxes 25 The Esplanade                                        98   
Studio District                                                        38   
The Annex,North Midtown,Yorkville                                      21   
The Beaches                                                             4   
The Beaches West,India Bazaar                                          21   
The Danforth West,Riverdale                                            42   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Adelaide,King,Richmond                                                 100   
Berczy Park                                                             58   
Brockton,Exhibition Place,Parkdale Village                              22   
Business Reply Mail Processing Centre 969 Eastern                       18   
CN Tower,Bathurst Quay,Island airport,Harbourfr...                      16   
Cabbagetown,St. James Town                                              44   
Central Bay Street                                                      83   
Chinatown,Grange Park,Kensington Market                                 95   
Christie                                                                17   
Church and Wellesley                                                    85   
Commerce Court,Victoria Hotel      

#### Let's find out how many unique categories can be curated from all the returned venues


In [65]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 230 uniques categories.


## 3. Analyze Each Neighborhood

For analysis using k means clustering we have to convert these venues categories into numerical form using dummies(one hot encoding)

In [68]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

toronto_onehot.head()

Yoga Studio  Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0            0                  0        0                   0             0   
1            0                  0        0                   0             0   
2            0                  0        0                   0             0   
3            0                  0        0                   0             0   
4            0                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Argentinian Restaurant  Art Gallery  \
0             0         0                       0            0   
1             0         0                       0            0   
2             0         0                       0            0   
3             0         0                       0            0   
4             0         0                       0            0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Workshop  \
0                    0                 0                   0              0   
1                    0                 0                   0              0   
2                    0                 0                   0              0   
3                    0                 0                   0              0   
4                    0                 0                   0              0   

   BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  Baseball Stadium  \
0          0           0           0       0     0    0                 0   
1          0           0           0       0     0    0                 0   
2          0           0           0       0     0    0                 0   
3          0           0           0       0     0    0                 0   
4          0           0           0       0     0    0                 0   

   Basketball Stadium  Beach  Bed & Breakfast  Beer Bar  Beer Store  \
0                   0      0                0         0           0   
1                   0      0                0         0           0   
2                   0      0                0         0           0   
3                   0      0                0         0           0   
4                   0      0                0         0           0   

   Belgian Restaurant  Bistro  Boat or Ferry  Bookstore  Boutique  \
0                   0       0              0          0         0   
1                   0       0              0          0         0   
2                   0       0              0          0         0   
3                   0       0              0          0         0   
4                   0       0              0          0         0   

   Brazilian Restaurant  Breakfast Spot  Brewery  Bubble Tea Shop  Building  \
0                     0               0        0                0         0   
1                     0               0        0                0         0   
2                     0               0        0                0         0   
3                     0               0        0                0         0   
4                     0               0        0                0         0   

   Burger Joint  Burrito Place  Bus Line  Butcher  Café  \
0             0              0         0        0     0   
1             0              0         0        0     0   
2             0              0         0        0     0   
3             0              0         0        0     0   
4             0              0         0        0     0   

   Cajun / Creole Restaurant  Candy Store  Caribbean

Move "Neighborhood" column to first position.

In [83]:
# re ordering the columns
cols = ['Neighborhood' , 
             'Yoga Studio',
 'Afghan Restaurant',
 'Airport',
 'Airport Food Court',
 'Airport Gate',
 'Airport Lounge',
 'Airport Service',
 'Airport Terminal',
 'American Restaurant',
 'Antique Shop',
 'Aquarium',
 'Argentinian Restaurant',
 'Art Gallery',
 'Arts & Crafts Store',
 'Asian Restaurant',
 'Athletics & Sports',
 'Auto Workshop',
 'BBQ Joint',
 'Baby Store',
 'Bagel Shop',
 'Bakery',
 'Bank',
 'Bar',
 'Baseball Stadium',
 'Basketball Stadium',
 'Beach',
 'Bed & Breakfast',
 'Beer Bar',
 'Beer Store',
 'Belgian Restaurant',
 'Bistro',
 'Boat or Ferry',
 'Bookstore',
 'Boutique',
 'Brazilian Restaurant',
 'Breakfast Spot',
 'Brewery',
 'Bubble Tea Shop',
 'Building',
 'Burger Joint',
 'Burrito Place',
 'Bus Line',
 'Butcher',
 'Café',
 'Cajun / Creole Restaurant',
 'Candy Store',
 'Caribbean Restaurant',
 'Cheese Shop',
 'Chinese Restaurant',
 'Chocolate Shop',
 'Church',
 'Climbing Gym',
 'Clothing Store',
 'Cocktail Bar',
 'Coffee Shop',
 'College Arts Building',
 'College Gym',
 'College Rec Center',
 'Colombian Restaurant',
 'Comfort Food Restaurant',
 'Comic Shop',
 'Concert Hall',
 'Convenience Store',
 'Cosmetics Shop',
 'Coworking Space',
 'Creperie',
 'Cuban Restaurant',
 'Cupcake Shop',
 'Dance Studio',
 'Deli / Bodega',
 'Department Store',
 'Dessert Shop',
 'Dim Sum Restaurant',
 'Diner',
 'Discount Store',
 'Dog Run',
 'Doner Restaurant',
 'Donut Shop',
 'Dumpling Restaurant',
 'Eastern European Restaurant',
 'Electronics Store',
 'Ethiopian Restaurant',
 'Event Space',
 'Falafel Restaurant',
 'Farmers Market',
 'Fast Food Restaurant',
 'Filipino Restaurant',
 'Fish & Chips Shop',
 'Fish Market',
 'Flea Market',
 'Food',
 'Food & Drink Shop',
 'Food Court',
 'Food Truck',
 'Fountain',
 'French Restaurant',
 'Fried Chicken Joint',
 'Fruit & Vegetable Store',
 'Furniture / Home Store',
 'Gaming Cafe',
 'Garden',
 'Garden Center',
 'Gastropub',
 'Gay Bar',
 'General Entertainment',
 'General Travel',
 'German Restaurant',
 'Gift Shop',
 'Gluten-free Restaurant',
 'Gourmet Shop',
 'Greek Restaurant',
 'Grocery Store',
 'Gym',
 'Gym / Fitness Center',
 'Harbor / Marina',
 'Health & Beauty Service',
 'Health Food Store',
 'Historic Site',
 'History Museum',
 'Hobby Shop',
 'Hospital',
 'Hostel',
 'Hotel',
 'Hotel Bar',
 'Hotpot Restaurant',
 'Ice Cream Shop',
 'Indian Restaurant',
 'Indie Movie Theater',
 'Intersection',
 'Irish Pub',
 'Italian Restaurant',
 'Japanese Restaurant',
 'Jazz Club',
 'Jewelry Store',
 'Juice Bar',
 'Korean Restaurant',
 'Lake',
 'Latin American Restaurant',
 'Light Rail Station',
 'Lingerie Store',
 'Liquor Store',
 'Lounge',
 'Mac & Cheese Joint',
 'Malay Restaurant',
 'Market',
 'Massage Studio',
 'Mediterranean Restaurant',
 "Men's Store",
 'Mexican Restaurant',
 'Middle Eastern Restaurant',
 'Miscellaneous Shop',
 'Modern European Restaurant',
 'Molecular Gastronomy Restaurant',
 'Monument / Landmark',
 'Movie Theater',
 'Museum',
 'Music Venue',
 'New American Restaurant',
 'Nightclub',
 'Noodle House',
 'Office',
 'Opera House',
 'Optical Shop',
 'Organic Grocery',
 'Other Great Outdoors',
 'Park',
 'Performing Arts Venue',
 'Persian Restaurant',
 'Pet Store',
 'Pharmacy',
 'Pizza Place',
 'Plane',
 'Playground',
 'Plaza',
 'Poke Place',
 'Portuguese Restaurant',
 'Poutine Place',
 'Pub',
 'Ramen Restaurant',
 'Record Shop',
 'Rental Car Location',
 'Restaurant',
 'Roof Deck',
 'Sake Bar',
 'Salad Place',
 'Salon / Barbershop',
 'Sandwich Place',
 'Scenic Lookout',
 'Sculpture Garden',
 'Seafood Restaurant',
 'Shoe Store',
 'Shopping Mall',
 'Skate Park',
 'Skating Rink',
 'Smoke Shop',
 'Smoothie Shop',
 'Snack Place',
 'Soup Place',
 'Southern / Soul Food Restaurant',
 'Spa',
 'Speakeasy',
 'Sporting Goods Shop',
 'Sports Bar',
 'Stadium',
 'Stationery Store',
 'Steakhouse',
 'Strip Club',
 'Supermarket',
 'Sushi Restaurant',
 'Swim School',
 'Taco Place',
 'Tailor Shop',
 'Taiwanese Restaurant',
 'Tanning Salon',
 'Tea Room',
 'Tennis Court',
 'Thai Restaurant',
 'Theater',
 'Theme Restaurant',
 'Thrift / Vintage Store',
 'Toy / Game Store',
 'Trail',
 'Train Station',
 'Vegetarian / Vegan Restaurant',
 'Video Game Store',
 'Vietnamese Restaurant',
 'Wine Bar',
 'Wine Shop',
 'Wings Joint']

In [85]:
toronto_onehot = toronto_onehot[cols]
toronto_onehot.head()

Neighborhood  Yoga Studio  Afghan Restaurant  Airport  \
0                  The Beaches            0                  0        0   
1                  The Beaches            0                  0        0   
2                  The Beaches            0                  0        0   
3                  The Beaches            0                  0        0   
4  The Danforth West,Riverdale            0                  0        0   

   Airport Food Court  Airport Gate  Airport Lounge  Airport Service  \
0                   0             0               0                0   
1                   0             0               0                0   
2                   0             0               0                0   
3                   0             0               0                0   
4                   0             0               0                0   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  \
0                 0                    0             0         0   
1                 0                    0             0         0   
2                 0                    0             0         0   
3                 0                    0             0         0   
4                 0                    0             0         0   

   Argentinian Restaurant  Art Gallery  Arts & Crafts Store  Asian Restaurant  \
0                       0            0                    0                 0   
1                       0            0                    0                 0   
2                       0            0                    0                 0   
3                       0            0                    0                 0   
4                       0            0                    0                 0   

   Athletics & Sports  Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  \
0                   0              0          0           0           0   
1                   0              0          0           0           0   
2                   0              0          0           0           0   
3                   0              0          0           0           0   
4                   0              0          0           0           0   

   Bakery  Bank  Bar  Baseball Stadium  Basketball Stadium  Beach  \
0       0     0    0                 0                   0      0   
1       0     0    0                 0                   0      0   
2       0     0    0                 0                   0      0   
3       0     0    0                 0                   0      0   
4       0     0    0                 0                   0      0   

   Bed & Breakfast  Beer Bar  Beer Store  Belgian Restaurant  Bistro  \
0                0         0           0                   0       0   
1                0         0           0                   0       0   
2                0         0           0                   0       0   
3                0         0           0                   0       0   
4                0         0           0                   0       0   

   Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  \
0              0          0         0                     0               0   
1              0          0         0                     0               0   
2              0          0         0                     0               0   
3              0          0         0                     0               0   
4              0          0         0                     0               0   

   Brewery  Bubble Tea Shop  Building  Burger Joint  Burrito Place  Bus Line  \
0        0                0         0             0              0         0   
1        0                0         0             0              0         0   
2        0                0         0             0              0         0   
3        0                0         0             0              0         0   
4        0                0         0             0             

And let's examine the new dataframe size.

In [87]:
toronto_onehot.shape

(1698, 230)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [88]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Yoga Studio  \
0                              Adelaide,King,Richmond     0.000000   
1                                         Berczy Park     0.000000   
2          Brockton,Exhibition Place,Parkdale Village     0.045455   
3   Business Reply Mail Processing Centre 969 Eastern     0.055556   
4   CN Tower,Bathurst Quay,Island airport,Harbourf...     0.000000   
5                          Cabbagetown,St. James Town     0.000000   
6                                  Central Bay Street     0.012048   
7             Chinatown,Grange Park,Kensington Market     0.000000   
8                                            Christie     0.000000   
9                                Church and Wellesley     0.011765   
10                      Commerce Court,Victoria Hotel     0.000000   
11                                         Davisville     0.000000   
12                                   Davisville North     0.000000   
13  Deer Park,Forest Hill SE,Rathnelly,South Hill,...     0.000000   
14            Design Exchange,Toronto Dominion Centre     0.000000   
15                        Dovercourt Village,Dufferin     0.000000   
16              First Canadian Place,Underground city     0.000000   
17                 Forest Hill North,Forest Hill West     0.000000   
18                      Harbord,University of Toronto     0.000000   
19                                       Harbourfront     0.020833   
20    Harbourfront East,Toronto Islands,Union Station     0.000000   
21                       High Park,The Junction South     0.000000   
22                                      Lawrence Park     0.000000   
23                            Little Portugal,Trinity     0.015625   
24                         Moore Park,Summerhill East     0.000000   
25                                 North Toronto West     0.047619   
26                              Parkdale,Roncesvalles     0.000000   
27                                           Rosedale     0.000000   
28                                           Roselawn     0.000000   
29                                  Runnymede,Swansea     0.000000   
30                            Ryerson,Garden District     0.000000   
31                                     St. James Town     0.000000   
32                    Stn A PO Boxes 25 The Esplanade     0.000000   
33                                    Studio District     0.026316   
34                  The Annex,North Midtown,Yorkville     0.000000   
35                                        The Beaches     0.000000   
36                      The Beaches West,India Bazaar     0.000000   
37                        The Danforth West,Riverdale     0.023810   

    Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0            0.000000   0.0000              0.0000        0.0000   
1            0.000000   0.0000              0.0000        0.0000   
2            0.000000   0.0000              0.0000        0.0000   
3            0.000000   0.0000              0.0000        0.0000   
4            0.000000   0.0625              0.0625        0.0625   
5            0.000000   0.0000              0.0000        0.0000   
6            0.000000   0.0000              0.0000        0.0000   
7            0.000000   0.0000              0.0000        0.0000   
8            0.000000   0.0000              0.0000        0.0000   
9            0.011765   0.0000              0.0000        0.0000   
10           0.000000   0.0000              0.0000        0.0000   
11           0.000000   0.0000              0.0000        0.0000   
12           0.000000   0.0000              0.0000        0.0000   
13           0.000000   0.0000              0.0000        0.0000   
14           0.000000   0.0000              0.0000        0.0000   
15           0.000000   0.0000              0.0000        0.0000   
16           0.000000   0.0000              0.0000        0.0000   
17           0.000000   0.0000              0.0000        0.0000   
18           0.000

#### Let's confirm the new size

In [89]:
toronto_grouped.shape

(38, 230)

#### Let's print each neighborhood along with the top 5 most common venues

In [90]:
num_top_venues = 5

for hood in toronto_grouped['Neighborhood']:
    print("----"+hood+"----")
    temp = toronto_grouped[toronto_grouped['Neighborhood'] == hood].T.reset_index()
    temp.columns = ['venue','freq']
    temp = temp.iloc[1:]
    temp['freq'] = temp['freq'].astype(float)
    temp = temp.round({'freq': 2})
    print(temp.sort_values('freq', ascending=False).reset_index(drop=True).head(num_top_venues))
    print('\n')

----Adelaide,King,Richmond----
             venue  freq
0      Coffee Shop  0.07
1             Café  0.05
2  Thai Restaurant  0.04
3              Bar  0.04
4       Steakhouse  0.04


----Berczy Park----
            venue  freq
0     Coffee Shop  0.09
1          Bakery  0.05
2  Farmers Market  0.03
3    Cocktail Bar  0.03
4     Cheese Shop  0.03


----Brockton,Exhibition Place,Parkdale Village----
                venue  freq
0      Breakfast Spot  0.09
1                Café  0.09
2         Coffee Shop  0.09
3         Yoga Studio  0.05
4  Italian Restaurant  0.05


----Business Reply Mail Processing Centre 969 Eastern----
                venue  freq
0  Light Rail Station  0.11
1         Yoga Studio  0.06
2       Auto Workshop  0.06
3          Smoke Shop  0.06
4         Pizza Place  0.06


----CN Tower,Bathurst Quay,Island airport,Harbourfront West,King and Spadina,Railway Lands,South Niagara----
              venue  freq
0   Airport Service  0.19
1    Airport Lounge  0.12
2  Airport Term

#### Let's put that into a *pandas* dataframe

First, let's write a function to sort the venues in descending order.

In [91]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

Now let's create the new dataframe and display the top 10 venues for each neighborhood.

In [94]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                             Adelaide,King,Richmond           Coffee Shop   
1                                        Berczy Park           Coffee Shop   
2         Brockton,Exhibition Place,Parkdale Village        Breakfast Spot   
3  Business Reply Mail Processing Centre 969 Eastern    Light Rail Station   
4  CN Tower,Bathurst Quay,Island airport,Harbourf...       Airport Service   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                  Café       Thai Restaurant            Steakhouse   
1                Bakery              Beer Bar          Cocktail Bar   
2                  Café           Coffee Shop           Yoga Studio   
3           Yoga Studio         Auto Workshop            Comic Shop   
4      Airport Terminal        Airport Lounge       Harbor / Marina   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0                   Bar      Sushi Restaurant   American Restaurant   
1        Farmers Market            Steakhouse    Seafood Restaurant   
2               Stadium                   Bar    Italian Restaurant   
3           Pizza Place            Restaurant               Butcher   
4           Coffee Shop      Sculpture Garden         Boat or Ferry   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0                Bakery            Restaurant         Breakfast Spot  
1           Cheese Shop                  Café         Clothing Store  
2                Bakery          Intersection           Climbing Gym  
3         Burrito Place               Brewery             Skate Park  
4                   Bar                 Plane           Airport Gate

## 4. Cluster Neighborhoods


Run *k*-means to cluster the neighborhood into 5 clusters.

In [100]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10] 

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [102]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head() # check the last columns!

PostalCode          Borough                   Neighborhood   Latitude  \
0        M4E     East Toronto                    The Beaches  43.676357   
1        M4K     East Toronto    The Danforth West,Riverdale  43.679557   
2        M4L     East Toronto  The Beaches West,India Bazaar  43.668999   
3        M4M     East Toronto                Studio District  43.659526   
4        M4N  Central Toronto                  Lawrence Park  43.728020   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.293031               0     Health Food Store                 Trail   
1 -79.352188               0      Greek Restaurant           Coffee Shop   
2 -79.315572               0                  Park        Sandwich Place   
3 -79.340923               0                  Café           Coffee Shop   
4 -79.388790               4              Bus Line                  Park   

  3rd Most Common Venue 4th Most Common Venue   5th Most Common Venue  \
0                   Pub           Wings Joint        Department Store   
1        Ice Cream Shop    Italian Restaurant  Furniture / Home Store   
2             Pet Store                   Pub            Liquor Store   
3                Bakery    Italian Restaurant     American Restaurant   
4           Swim School           Wings Joint    Ethiopian Restaurant   

  6th Most Common Venue        7th Most Common Venue  \
0  Ethiopian Restaurant            Electronics Store   
1            Restaurant               Cosmetics Shop   
2    Light Rail Station                 Burger Joint   
3     Convenience Store           Seafood Restaurant   
4     Electronics Store  Eastern European Restaurant   

         8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
0  Eastern European Restaurant   Dumpling Restaurant             Donut Shop  
1                    Bookstore               Brewery        Bubble Tea Shop  
2                Burrito Place  Fast Food Restaurant     Italian Restaurant  
3                      Brewery        Sandwich Place            Cheese Shop  
4          Dumpling Restaurant            Donut Shop       Doner Restaurant

In [139]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11.2)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]


# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(' Cluster ' + str(cluster + 1) + str(poi), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

## 5. Examine Clusters


#### Cluster 1

just making a seprate data frames for each cluster which will help us in further analysis

In [152]:
clus1 = toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]
clus1.head()

Borough  Cluster Labels 1st Most Common Venue  \
0     East Toronto               0     Health Food Store   
1     East Toronto               0      Greek Restaurant   
2     East Toronto               0                  Park   
3     East Toronto               0                  Café   
5  Central Toronto               0        Breakfast Spot   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                 Trail                   Pub           Wings Joint   
1           Coffee Shop        Ice Cream Shop    Italian Restaurant   
2        Sandwich Place             Pet Store                   Pub   
3           Coffee Shop                Bakery    Italian Restaurant   
5        Clothing Store     Food & Drink Shop                 Hotel   

    5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0        Department Store  Ethiopian Restaurant     Electronics Store   
1  Furniture / Home Store            Restaurant        Cosmetics Shop   
2            Liquor Store    Light Rail Station          Burger Joint   
3     American Restaurant     Convenience Store    Seafood Restaurant   
5       Convenience Store        Sandwich Place                  Park   

         8th Most Common Venue        9th Most Common Venue  \
0  Eastern European Restaurant          Dumpling Restaurant   
1                    Bookstore                      Brewery   
2                Burrito Place         Fast Food Restaurant   
3                      Brewery               Sandwich Place   
5                          Gym  Eastern European Restaurant   

  10th Most Common Venue  
0             Donut Shop  
1        Bubble Tea Shop  
2     Italian Restaurant  
3            Cheese Shop  
5    Dumpling Restaurant

In [240]:
clus1.groupby("Borough")["1st Most Common Venue"].value_counts()

Borough           1st Most Common Venue
Central Toronto   Breakfast Spot            1
                  Café                      1
                  Clothing Store            1
                  Coffee Shop               1
                  Pizza Place               1
Downtown Toronto  Coffee Shop              13
                  Café                      2
                  Airport Service           1
                  Grocery Store             1
East Toronto      Café                      1
                  Greek Restaurant          1
                  Health Food Store         1
                  Light Rail Station        1
                  Park                      1
West Toronto      Bar                       1
                  Breakfast Spot            1
                  Café                      1
                  Gift Shop                 1
                  Pharmacy                  1
                  Thai Restaurant           1
Name: 1st Most Common Venue, dtype: int6

In [174]:
clus1.groupby("Borough")["1st Most Common Venue"].apply(lambda x: ','.join(x.unique()))

Borough
Central Toronto     Breakfast Spot,Clothing Store,Pizza Place,Coff...
Downtown Toronto       Coffee Shop,Café,Airport Service,Grocery Store
East Toronto        Health Food Store,Greek Restaurant,Park,Café,L...
West Toronto        Pharmacy,Bar,Breakfast Spot,Thai Restaurant,Gi...
Name: 1st Most Common Venue, dtype: object

In [222]:
df_tmp_clus1 = pd.DataFrame(clus1.groupby("Borough")["1st Most Common Venue"].apply(lambda x: ','.join(x.unique())).reset_index())
df_tmp_clus1.rename(columns={'Borough':'boroughs', '1st Most Common Venue': 'top_venues'},inplace=True)
df_tmp_clus1.loc[df_tmp_clus1.boroughs == "Central Toronto","top_venues"].values

array(['Breakfast Spot,Clothing Store,Pizza Place,Coffee Shop,Café'],
      dtype=object)

From above groupby we can see that cluster 1 has many coffeshops and is the busiest as it is downtown of toronto.

#### Cluster 2

In [116]:
clus2 = toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]
clus2.head()

Borough  Cluster Labels 1st Most Common Venue  \
8    Central Toronto               1            Playground   
10  Downtown Toronto               1                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
8                   Park          Tennis Court            Restaurant   
10            Playground                 Trail           Wings Joint   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
8            Wings Joint        Discount Store          Dessert Shop   
10      Department Store  Ethiopian Restaurant     Electronics Store   

          8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
8            Dim Sum Restaurant                 Diner       Doner Restaurant  
10  Eastern European Restaurant   Dumpling Restaurant             Donut Shop

In [225]:
df_tmp_clus2 = pd.DataFrame(clus2.groupby("Borough")["1st Most Common Venue"].apply(lambda x: ','.join(x.unique())).reset_index())
df_tmp_clus2.rename(columns={'Borough':'boroughs', '1st Most Common Venue': 'top_venues'},inplace=True)
df_tmp_clus2.loc[df_tmp_clus2.boroughs == "Central Toronto","top_venues"].values

array(['Playground'], dtype=object)

we can observe that cluster2 has just 2 points one from central and one from downtown toronto and is more focused on parks and play grounds. 

#### Cluster 3

In [117]:
clus3 = toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]
clus3.head()

Borough  Cluster Labels 1st Most Common Venue  \
22  Central Toronto               2                Garden   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
22           Wings Joint      Department Store           Event Space   

   5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
22  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
22   Dumpling Restaurant            Donut Shop       Doner Restaurant

In [226]:
df_tmp_clus3 = pd.DataFrame(clus3.groupby("Borough")["1st Most Common Venue"].apply(lambda x: ','.join(x.unique())).reset_index())
df_tmp_clus3.rename(columns={'Borough':'boroughs', '1st Most Common Venue': 'top_venues'},inplace=True)
df_tmp_clus3.loc[df_tmp_clus3.boroughs == "Central Toronto","top_venues"].values

array(['Garden'], dtype=object)

Cluster 3 has gardens and seems great chillout neighbourhood.

#### Cluster 4

In [118]:
clus4 = toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]
clus4.head()

Borough  Cluster Labels 1st Most Common Venue  \
23  Central Toronto               3              Bus Line   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
23         Jewelry Store                  Park      Sushi Restaurant   

   5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
23                 Trail    Dim Sum Restaurant                 Diner   

   8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
23        Discount Store               Dog Run            Wings Joint

In [227]:
df_tmp_clus4 = pd.DataFrame(clus4.groupby("Borough")["1st Most Common Venue"].apply(lambda x: ','.join(x.unique())).reset_index())
df_tmp_clus4.rename(columns={'Borough':'boroughs', '1st Most Common Venue': 'top_venues'},inplace=True)
df_tmp_clus4.loc[df_tmp_clus4.boroughs == "Central Toronto","top_venues"].values

array(['Bus Line'], dtype=object)

Cluster 4 is all about shopping and public transport relatable right?

#### Cluster 5

In [119]:
clus5 = toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[1] + list(range(5, toronto_merged.shape[1]))]]
clus5.head()

Borough  Cluster Labels 1st Most Common Venue  \
4  Central Toronto               4              Bus Line   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
4                  Park           Swim School           Wings Joint   

  5th Most Common Venue 6th Most Common Venue        7th Most Common Venue  \
4  Ethiopian Restaurant     Electronics Store  Eastern European Restaurant   

  8th Most Common Venue 9th Most Common Venue 10th Most Common Venue  
4   Dumpling Restaurant            Donut Shop       Doner Restaurant

In [235]:
df_tmp_clus5 = pd.DataFrame(clus5.groupby("Borough")["1st Most Common Venue"].apply(lambda x: ','.join(x.unique())).reset_index())
df_tmp_clus5.rename(columns={'Borough':'boroughs', '1st Most Common Venue': 'top_venues'},inplace=True)
df_tmp_clus5.loc[df_tmp_clus5.boroughs == "Central Toronto","top_venues"].values

array(['Bus Line'], dtype=object)

Cluster 5 is mix restaurants and stores also, we have public transit more focused here.

In [244]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11.2)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]


# add markers to the map
markers_colors = []
for lat, lon, poi, cluster, top_spot, borough  in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels'],
                                 toronto_merged['1st Most Common Venue'], toronto_merged['Borough']):
    # putting the extra options to explore in each cluster on the popups 
    if cluster == 0: 
        label = folium.Popup("<h6> <center>Coffee House </center></h6> \
                             <h6><b> {}</b> : {}</h6> \
                          <h6> <b> {} </b> : {} </h6> ".format(str(poi),str(top_spot),str(borough),str(df_tmp_clus1.loc[df_tmp_clus1.boroughs == borough,"top_venues"].values)),parse_html=False)
    elif cluster == 1:
        label = folium.Popup("<center>Parks and Playgrounds </center>\
                             <h6><b> {}</b> : {}</h6> \
                          <h6> <b> {} </b> : {} </h6> ".format(str(poi),str(top_spot),str(borough),str(df_tmp_clus2.loc[df_tmp_clus2.boroughs == borough,"top_venues"].values)),parse_html=False)
    
        
    elif cluster == 2:
        label = folium.Popup("<h6> <center>Gardens </center> </h6> \
                             <h6><b> {}</b> : {}</h6> \
                          <h6> <b> {} </b> : {} </h6> ".format(str(poi),str(top_spot),str(borough),str(df_tmp_clus3.loc[df_tmp_clus3.boroughs == borough,"top_venues"].values)),parse_html=False)
        
    elif cluster == 3:
        label = folium.Popup("<h6> <center> Shoping and Restaurants </center> </h6> \
                             <h6><b> {}</b> : {}</h6> \
                          <h6> <b> {} </b> : {} </h6> ".format(str(poi),str(top_spot),str(borough),str(df_tmp_clus4.loc[df_tmp_clus4.boroughs == borough,"top_venues"].values)),parse_html=False)
    
    elif cluster == 4:
        label = folium.Popup("<h6> <center>Stores and transport</center> </h6> \
                             <h6><b> {}</b> : {}</h6> \
                          <h6> <b> {} </b> : {} </h6> ".format(str(poi),str(top_spot),str(borough),str(df_tmp_clus5.loc[df_tmp_clus5.boroughs == borough,"top_venues"].values)),parse_html=False)
        
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

#### from above map we can see that our neighborhoods analysis shows the results in popup with trending venues from foursquare api. which lets us know which is first and foremost place to go will be the one whcich is famous in neighborhood and other things trending in that Borough are mentioned below the line. 
#### we have sucesfully analysed neighborhoods in toronto which let us know which are the places to go for coffe in downtwon or where to go for some relaxation( like cluster 3 having parks and playgrounds ).
if some one is looking to buy a house on toronto then we can se we have 5 options 
1. if he wants all the resources and office near home he can buy house in cluser 1.
2. if he wants rexalxed non busy envoirnment then he should go for 2nd cluster area.
3. if he wants gardens nearby go for cluster 3.
4. if he wants more of public transport as well as shopping and stores nearby home he should go for 4th or 5th cluster aera.

These sugessions can be further broken down as per boroughs.